sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala

```diff
@@ -1207,6 +1207,17 @@ class Analyzer(
                 val alias = Alias(ae, ae.toString)()
                 aggregateExpressions += alias
                 alias.toAttribute
+              // Replacing [[NamedExpression]] causes the error on [[Grouping]] because the
+              // grouping column will be new attribute created by adding additional [[Alias]].
+              // So we can't find the grouping column and replace it in the rule
+              // [[ResolveGroupingAnalytics]].
+              case ne: NamedExpression => ne
+              // Grouping functions are handled in the rule [[ResolveGroupingAnalytics]].
+              case e: Expression if grouping.exists(_.semanticEquals(e)) &&
+                  !ResolveGroupingAnalytics.hasGroupingFunction(e) =>
```

Comment:
You mean if we don't add this check? Yes. As the grouping function is pushed down to Aggregation, the rule `ResolveGroupingAnalytics` can't perform the replacement of grouping function.